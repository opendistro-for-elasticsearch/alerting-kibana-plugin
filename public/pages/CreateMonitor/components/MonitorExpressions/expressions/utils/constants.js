/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const POPOVER_STYLE = { zIndex: '200' };
export const EXPRESSION_STYLE = { padding: '20px', whiteSpace: 'nowrap' };
export const Expressions = {
  THRESHOLD: 'THRESHOLD',
  WHEN: 'WHEN',
  OF_FIELD: 'OF_FIELD',
  OVER: 'OVER',
  FOR_THE_LAST: 'FOR_THE_LAST',
  WHERE: 'WHERE',
};
export const NUMBER_TYPES = [
  'long',
  'integer',
  'short',
  'byte',
  'double',
  'float',
  'half_float',
  'scaled_float',
];
export const UNITS_OF_TIME = [
  { value: 'm', text: 'minute(s)' },
  { value: 'h', text: 'hour(s)' },
  { value: 'd', text: 'day(s)' },
];

export const WHERE_BOOLEAN_FILTERS = [
  { text: 'Select value', value: '' },
  { text: 'True', value: true },
  { text: 'False', value: false },
];

export const OPERATORS_MAP = {
  IS: 'is',
  IS_NOT: 'is_not',
  IS_NULL: 'is_null',
  IS_NOT_NULL: 'is_not_null',
  IS_GREATER: 'is_greater',
  IS_GREATER_EQUAL: 'is_greater_equal',
  IS_LESS: 'is_lesser',
  IS_LESS_EQUAL: 'is_lesser_equal',
  IS_LESS_EQUAL: 'is_lesser_equal',
  STARTS_WITH: 'starts_with',
  ENDS_WITH: 'ends_with',
  CONTAINS: 'contains',
  NOT_CONTAINS: 'does_not_contains',
  IN_RANGE: 'in_range',
  NOT_IN_RANGE: 'not_in_range',
};

export const COMPARISON_OPERATORS = [
  { text: 'is', value: OPERATORS_MAP.IS, dataTypes: ['number', 'text', 'keyword', 'boolean'] },
  {
    text: 'is not',
    value: OPERATORS_MAP.IS_NOT,
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  {
    text: 'is null',
    value: OPERATORS_MAP.IS_NULL,
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  {
    text: 'is not null',
    value: OPERATORS_MAP.IS_NOT_NULL,
    dataTypes: ['number', 'text', 'keyword'],
  },
  { text: 'is greater than', value: OPERATORS_MAP.IS_GREATER, dataTypes: ['number'] },
  { text: 'is greater than equal', value: OPERATORS_MAP.IS_GREATER_EQUAL, dataTypes: ['number'] },
  { text: 'is less than', value: OPERATORS_MAP.IS_LESS, dataTypes: ['number'] },
  { text: 'is less than equal', value: OPERATORS_MAP.IS_LESS_EQUAL, dataTypes: ['number'] },
  { text: 'is in range', value: OPERATORS_MAP.IN_RANGE, dataTypes: ['number'] },
  { text: 'is not in range', value: OPERATORS_MAP.NOT_IN_RANGE, dataTypes: ['number'] },
  { text: 'starts with', value: OPERATORS_MAP.STARTS_WITH, dataTypes: ['text', 'keyword'] },
  { text: 'ends with', value: OPERATORS_MAP.ENDS_WITH, dataTypes: ['text', 'keyword'] },
  { text: 'contains', value: OPERATORS_MAP.CONTAINS, dataTypes: ['text', 'keyword'] },
  { text: 'does not contains', value: OPERATORS_MAP.NOT_CONTAINS, dataTypes: ['text'] },
];
export const OVER_TYPES = [{ value: 'all documents', text: 'all documents' }];

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];
