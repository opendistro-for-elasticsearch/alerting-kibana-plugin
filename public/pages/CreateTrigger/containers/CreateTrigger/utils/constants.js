/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const TRIGGER_TYPE = {
  AD: 'anomaly_detector_trigger',
  ALERT_TRIGGER: 'alerting_trigger',
};

export const FORMIK_INITIAL_VALUES = {
  name: '',
  severity: '1',
  minTimeBetweenExecutions: null,
  rollingWindowSize: null,
  script: {
    lang: 'painless',
    source: `ctx.results[0].hits.total.value > 0`,
  },
  thresholdValue: 10000,
  thresholdEnum: 'ABOVE',
  anomalyDetector: {
    triggerType: TRIGGER_TYPE.AD,
    anomalyGradeThresholdValue: 0.7,
    anomalyGradeThresholdEnum: 'ABOVE',
    anomalyConfidenceThresholdValue: 0.7,
    anomalyConfidenceThresholdEnum: 'ABOVE',
  },
  actions: undefined,
};

export const HITS_TOTAL_RESULTS_PATH = 'ctx.results[0].hits.total.value';
export const AGGREGATION_RESULTS_PATH = 'ctx.results[0].aggregations.when.value';
export const ANOMALY_GRADE_RESULT_PATH = 'ctx.results[0].aggregations.max_anomaly_grade.value';
export const ANOMALY_CONFIDENCE_RESULT_PATH = 'ctx.results[0].hits.hits[0]._source.confidence';
