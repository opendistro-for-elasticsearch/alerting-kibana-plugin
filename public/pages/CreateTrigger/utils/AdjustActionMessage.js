/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

/* Mustache templates have different implementation in different programming languages,
 * thus the way of rendering may vary except for those are defined in the official manual.
 * Alerting Kibana plugin uses 'Mustache.js' while Alerting plugin uses 'Mustache.java' to render templates.
 * The below adjustment for Mustache templates aims to clean up differences between the two render engines
 * for the user when using Alerting Kibana plugin.
 */

import _ from 'lodash';

export const adjustMessageForPreview = (str) =>
  str.replace(/({{\s*[\w\.]+\.)size(\s*}})/g, '$1length$2'); // replace "size" by "length"

export const adjustMessageForBackend = (str) =>
  str
    .replace(/({{)\s*((#|\/|\^)\w[\w\.]*\s*}})/g, '$1$2') // remove leading spaces of sections
    .replace(/({{\s*[\w\.]+\.)length(\s*}})/g, '$1size$2'); // replace "length" by "size"

export const adjustMessageByTrigger = (trigger) => {
  const actions = trigger.actions;
  if (actions && actions.length > 0) {
    return actions.map((action) => adjustMessageByAction(action));
  }
};

export const adjustMessageByAction = (action) => {
  const source = action.message_template.source;
  _.set(action, 'message_template.source', adjustMessageForBackend(source));
};
