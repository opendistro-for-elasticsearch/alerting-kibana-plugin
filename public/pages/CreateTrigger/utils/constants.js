/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

const DEFAULT_MESSAGE_SOURCE = `
Monitor {{ctx.monitor.name}} just entered alert status. Please investigate the issue.
- Trigger: {{ctx.trigger.name}}
- Severity: {{ctx.trigger.severity}}
- Period start: {{ctx.periodStart}}
- Period end: {{ctx.periodEnd}}
`.trim();

export const FORMIK_INITIAL_ACTION_VALUES = {
  name: '',
  destination_id: '',
  subject_template: {
    lang: 'mustache',
    source: '',
  },
  message_template: {
    lang: 'mustache',
    source: DEFAULT_MESSAGE_SOURCE,
  },
  throttle_enabled: false,
  throttle: {
    value: 10,
    unit: 'MINUTES' // throttle unit only supports MINUTES currently, no UI element mapped
  }
};

export const DEFAULT_ACTION_TYPE = 'slack';
