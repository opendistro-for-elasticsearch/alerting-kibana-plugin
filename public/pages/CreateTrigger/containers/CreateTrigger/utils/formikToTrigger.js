/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import {
  AGGREGATION_RESULTS_PATH,
  HITS_TOTAL_RESULTS_PATH,
  TRIGGER_TYPE,
  ANOMALY_GRADE_RESULT_PATH,
  ANOMALY_CONFIDENCE_RESULT_PATH,
} from './constants';
import { SEARCH_TYPE } from '../../../../../utils/constants';

export function formikToTrigger(values, monitorUiMetadata = {}) {
  const condition = formikToCondition(values, monitorUiMetadata);
  const actions = formikToAction(values);
  return {
    id: values.id,
    name: values.name,
    severity: values.severity,
    condition,
    actions: actions,
    min_time_between_executions: values.minTimeBetweenExecutions,
    rolling_window_size: values.rollingWindowSize,
  };
}

export function formikToAction(values) {
  const actions = values.actions;
  if (actions && actions.length > 0) {
    return actions.map(action => {
      if (!action.throttle_enabled) return _.omit(action, ['throttle']);
      return action;
    });
  }
  return actions;
}

export function formikToTriggerUiMetadata(values, monitorUiMetadata) {
  const { anomalyDetector, thresholdEnum, thresholdValue } = values;
  const searchType = _.get(monitorUiMetadata, 'search.searchType', 'query');
  let triggerMetadata = { value: thresholdValue, enum: thresholdEnum };
  //Store AD values only if AD trigger.
  if (searchType === SEARCH_TYPE.AD && anomalyDetector.triggerType === TRIGGER_TYPE.AD) {
    triggerMetadata.adTriggerMetadata = {
      triggerType: anomalyDetector.triggerType,
      anomalyGrade: {
        value: anomalyDetector.anomalyGradeThresholdValue,
        enum: anomalyDetector.anomalyGradeThresholdEnum,
      },
      anomalyConfidence: {
        value: anomalyDetector.anomalyConfidenceThresholdValue,
        enum: anomalyDetector.anomalyConfidenceThresholdEnum,
      },
    };
  }
  return { [values.name]: triggerMetadata };
}

export function formikToCondition(values, monitorUiMetadata = {}) {
  const { thresholdValue, thresholdEnum } = values;
  const searchType = _.get(monitorUiMetadata, 'search.searchType', 'query');
  const aggregationType = _.get(monitorUiMetadata, 'search.aggregationType', 'count');

  if (searchType === SEARCH_TYPE.QUERY) return { script: values.script };
  if (searchType === SEARCH_TYPE.AD) return getADCondition(values);
  const isCount = aggregationType === 'count';
  const resultsPath = getResultsPath(isCount);
  const operator = getOperator(thresholdEnum);
  return getCondition(resultsPath, operator, thresholdValue, isCount);
}

export function getADCondition(values) {
  const { anomalyDetector } = values;
  if (anomalyDetector.triggerType === TRIGGER_TYPE.AD) {
    const anomalyGradeOperator = getOperator(anomalyDetector.anomalyGradeThresholdEnum);
    const anomalyConfidenceOperator = getOperator(anomalyDetector.anomalyConfidenceThresholdEnum);
    return {
      script: {
        lang: 'painless',
        //TODO:: Check null check is required or not confirm with backend
        source: `return ${ANOMALY_GRADE_RESULT_PATH} != null && ${ANOMALY_GRADE_RESULT_PATH} ${anomalyGradeOperator} ${anomalyDetector.anomalyGradeThresholdValue} && ${ANOMALY_CONFIDENCE_RESULT_PATH} ${anomalyConfidenceOperator} ${anomalyDetector.anomalyConfidenceThresholdValue}`,
      },
    };
  } else {
    return { script: values.script };
  }
}

export function getCondition(resultsPath, operator, value, isCount) {
  const baseSource = `${resultsPath} ${operator} ${value}`;
  return {
    script: {
      lang: 'painless',
      source: isCount ? baseSource : `return ${resultsPath} == null ? false : ${baseSource}`,
    },
  };
}

export function getResultsPath(isCount) {
  return isCount ? HITS_TOTAL_RESULTS_PATH : AGGREGATION_RESULTS_PATH;
}

export function getOperator(thresholdEnum) {
  return { ABOVE: '>', BELOW: '<', EXACTLY: '==' }[thresholdEnum];
}
