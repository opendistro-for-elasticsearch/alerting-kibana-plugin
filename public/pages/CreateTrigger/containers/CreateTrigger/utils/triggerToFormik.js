/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { FORMIK_INITIAL_VALUES, TRIGGER_TYPE } from './constants';

export function triggerToFormik(trigger, monitor) {
  const {
    id,
    name,
    severity,
    condition: { script },
    actions,
    min_time_between_executions: minTimeBetweenExecutions,
    rolling_window_size: rollingWindowSize,
  } = trigger;
  const thresholdEnum = _.get(
    monitor,
    `ui_metadata.triggers[${name}].enum`,
    FORMIK_INITIAL_VALUES.thresholdEnum
  );
  const thresholdValue = _.get(
    monitor,
    `ui_metadata.triggers[${name}].value`,
    FORMIK_INITIAL_VALUES.thresholdValue
  );
  const anomalyConfidenceThresholdValue = _.get(
    monitor,
    `ui_metadata.triggers[${name}].adTriggerMetadata.anomalyConfidence.value`,
    FORMIK_INITIAL_VALUES.anomalyDetector.anomalyConfidenceThresholdValue
  );
  const anomalyConfidenceThresholdEnum = _.get(
    monitor,
    `ui_metadata.triggers[${name}].adTriggerMetadata.anomalyConfidence.enum`,
    FORMIK_INITIAL_VALUES.anomalyDetector.anomalyConfidenceThresholdEnum
  );
  const anomalyGradeThresholdValue = _.get(
    monitor,
    `ui_metadata.triggers[${name}].adTriggerMetadata.anomalyGrade.value`,
    FORMIK_INITIAL_VALUES.anomalyDetector.anomalyGradeThresholdValue
  );
  const anomalyGradeThresholdEnum = _.get(
    monitor,
    `ui_metadata.triggers[${name}].adTriggerMetadata.anomalyGrade.enum`,
    FORMIK_INITIAL_VALUES.anomalyDetector.anomalyGradeThresholdEnum
  );
  const triggerType = _.get(monitor, `ui_metadata.triggers[${name}].adTriggerMetadata.triggerType`);
  return {
    ..._.cloneDeep(FORMIK_INITIAL_VALUES),
    id,
    name,
    severity,
    script,
    actions,
    minTimeBetweenExecutions,
    rollingWindowSize,
    thresholdEnum,
    thresholdValue,
    anomalyDetector: {
      /*If trigger type doesn't exist fallback to query trigger with following reasons
        1. User has changed monitory type from normal monitor to AD monitor.
        2. User has created / updated from API and visiting Kibana to do other operations.
      */
      triggerType: triggerType ? triggerType : TRIGGER_TYPE.ALERT_TRIGGER,
      anomalyGradeThresholdValue,
      anomalyGradeThresholdEnum,
      anomalyConfidenceThresholdValue,
      anomalyConfidenceThresholdEnum,
    },
  };
}
