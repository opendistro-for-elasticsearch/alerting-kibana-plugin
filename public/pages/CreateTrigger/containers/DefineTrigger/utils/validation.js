/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { FORMIK_INITIAL_TRIGGER_VALUES } from '../../CreateTrigger/utils/constants';

export const validateTriggerName = (triggers, triggerToEdit, fieldPath) => (value) => {
  if (!value) return 'Required';
  const nameExists = triggers.filter((trigger) => {
    const triggerId = _.get(
      trigger,
      `aggregation_trigger.id`,
      _.get(trigger, 'traditional_trigger.id')
    );
    const triggerName = _.get(
      trigger,
      'aggregation_trigger.name',
      _.get(trigger, `traditional_trigger.name`, FORMIK_INITIAL_TRIGGER_VALUES.name)
    );
    const triggerToEditId = _.get(triggerToEdit, `${fieldPath}.id`, triggerToEdit.id);
    return triggerToEditId !== triggerId && triggerName.toLowerCase() === value.toLowerCase();
  });
  if (nameExists.length > 0) {
    return 'Trigger name already used';
  }
  // TODO: character restrictions
  // TODO: character limits
};
