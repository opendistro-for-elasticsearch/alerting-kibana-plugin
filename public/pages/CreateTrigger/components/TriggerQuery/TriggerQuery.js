/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { Field } from 'formik';
import _ from 'lodash';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiCodeEditor,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
  EuiText,
} from '@elastic/eui';
import 'brace/mode/json';
import 'brace/mode/plain_text';
import 'brace/snippets/javascript';
import 'brace/ext/language_tools';
import { formikToTrigger } from '../../containers/CreateTrigger/utils/formikToTrigger';

export const getExecuteMessage = response => {
  if (!response) return 'No response';
  const triggerResults = _.get(response, 'trigger_results');
  if (!triggerResults) return 'No trigger results';
  const triggerId = Object.keys(triggerResults)[0];
  if (!triggerId) return 'No trigger results';
  const executeResults = _.get(triggerResults, `${triggerId}`);
  if (!executeResults) return 'No execute results';
  const { error, triggered } = executeResults;
  if (error) return `ERROR: ${error}`;
  return `${triggered}`;
};

const TriggerQuery = ({
  context,
  error,
  executeResponse,
  onRun,
  response,
  triggerValues,
  setFlyout,
  isDarkMode,
}) => {
  const trigger = { ...formikToTrigger(triggerValues), actions: [] };
  const formattedResponse = JSON.stringify(response, null, 4);
  return (
    <div style={{ padding: '0px 10px', marginTop: '0px' }}>
      <EuiFlexGroup direction="column">
        <EuiFlexItem>
          <EuiFormRow label="Extraction query response" fullWidth>
            <EuiCodeEditor
              mode="json"
              theme={isDarkMode ? 'sense-dark' : 'github'}
              width="100%"
              value={error || formattedResponse}
              readOnly
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem>
          <Field
            name="script.source"
            render={({
              field: { value },
              form: { errors, touched, setFieldValue, setFieldTouched },
            }) => (
              <EuiFormRow
                label={
                  <div>
                    <span>Trigger condition</span>
                    <EuiButtonEmpty
                      size="s"
                      onClick={() => {
                        setFlyout({ type: 'triggerCondition', payload: context });
                      }}
                    >
                      Info
                    </EuiButtonEmpty>
                  </div>
                }
                fullWidth
                isInvalid={
                  _.get(touched, 'script.source', false) && !!_.get(errors, 'script.source')
                }
                error={_.get(errors, 'script.source')}
              >
                <EuiCodeEditor
                  mode="plain_text"
                  theme={isDarkMode ? 'sense-dark' : 'github'}
                  height="300px"
                  width="100%"
                  onChange={source => {
                    setFieldValue('script.source', source);
                  }}
                  onBlur={() => setFieldTouched('script.source', true)}
                  value={value}
                />
              </EuiFormRow>
            )}
          />
        </EuiFlexItem>
      </EuiFlexGroup>

      <EuiSpacer size="s" />

      <EuiFlexGroup justifyContent="spaceBetween" alignItems="center">
        <EuiFlexItem grow={false}>
          <EuiText size="s">
            <strong>Trigger condition response: </strong>
            <span>{getExecuteMessage(executeResponse)}</span>
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton onClick={() => onRun([trigger])}>Run</EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
};

export default TriggerQuery;
