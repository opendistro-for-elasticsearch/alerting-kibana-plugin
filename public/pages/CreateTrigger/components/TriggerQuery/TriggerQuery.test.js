/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { shallow } from 'enzyme';
import { EuiButton } from '@elastic/eui';

import TriggerQuery, { getExecuteMessage } from './TriggerQuery';
import { FORMIK_INITIAL_VALUES } from '../../containers/CreateTrigger/utils/constants';
import { formikToTrigger } from '../../containers/CreateTrigger/utils/formikToTrigger';

const props = {
  context: {},
  executeResponse: null,
  onRun: jest.fn(),
  response: null,
  triggerValues: FORMIK_INITIAL_VALUES,
  setFlyout: jest.fn(),
};

describe('TriggerQuery', () => {
  test('renders', () => {
    const wrapper = shallow(<TriggerQuery {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('calls onRun when clicking Run', () => {
    const wrapper = shallow(<TriggerQuery {...props} />);
    const button = wrapper.find(EuiButton);
    button.simulate('click');
    expect(props.onRun).toHaveBeenCalled();
    expect(props.onRun).toHaveBeenCalledWith([
      { ...formikToTrigger(props.triggerValues), actions: [] },
    ]);
  });
});

describe('getExecuteMessage', () => {
  test('returns No Response for falsy response values', () => {
    const noReponse = 'No response';
    expect(getExecuteMessage(null)).toBe(noReponse);
    expect(getExecuteMessage('')).toBe(noReponse);
  });

  test('returns No trigger results for falsy trigger results values', () => {
    const noTriggerResults = 'No trigger results';
    expect(getExecuteMessage({})).toBe(noTriggerResults);
    expect(getExecuteMessage({ trigger_results: null })).toBe(noTriggerResults);
  });

  test('returns No trigger results for trigger_results with no trigger ids', () => {
    const noTriggerResults = 'No trigger results';
    expect(getExecuteMessage({ trigger_results: {} })).toBe(noTriggerResults);
  });

  test('returns No execute results for fasly trigger value on trigger_results', () => {
    const noTriggerResults = 'No execute results';
    expect(getExecuteMessage({ trigger_results: { trig_id: null } })).toBe(noTriggerResults);
  });

  test('returns triggered value', () => {
    expect(
      getExecuteMessage({ trigger_results: { trig_id: { triggered: true, error: null } } })
    ).toBe('true');
    expect(
      getExecuteMessage({ trigger_results: { trig_id: { triggered: false, error: null } } })
    ).toBe('false');
  });

  test('returns error message', () => {
    expect(
      getExecuteMessage({
        trigger_results: { trig_id: { triggered: true, error: 'RANDOM ERROR' } },
      })
    ).toBe('ERROR: RANDOM ERROR');
  });
});
