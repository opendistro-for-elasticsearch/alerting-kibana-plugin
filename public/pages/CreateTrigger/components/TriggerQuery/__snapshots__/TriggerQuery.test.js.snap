// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerQuery renders 1`] = `
<div
  style={
    Object {
      "marginTop": "0px",
      "padding": "0px 10px",
    }
  }
>
  <EuiFlexGroup
    direction="column"
  >
    <EuiFlexItem>
      <EuiFormRow
        describedByIds={Array []}
        fullWidth={true}
        hasEmptyLabelSpace={false}
        label="Extraction query response"
        labelType="label"
      >
        <EuiCodeEditor
          mode="json"
          readOnly={true}
          setOptions={Object {}}
          theme="github"
          value="null"
          width="100%"
        />
      </EuiFormRow>
    </EuiFlexItem>
    <EuiFlexItem>
      <C
        name="script.source"
        render={[Function]}
      />
    </EuiFlexItem>
    <EuiFlexItem
      grow={false}
    >
      <EuiFormRow
        describedByIds={Array []}
        fullWidth={true}
        hasEmptyLabelSpace={false}
        label="Trigger condition response:"
        labelType="label"
      >
        <EuiCodeEditor
          height="200px"
          readOnly={true}
          setOptions={Object {}}
          theme="github"
          value="No response"
          width="100%"
        />
      </EuiFormRow>
    </EuiFlexItem>
  </EuiFlexGroup>
  <EuiSpacer
    size="s"
  />
  <EuiFlexGroup
    alignItems="center"
    justifyContent="spaceBetween"
  >
    <EuiFlexItem
      grow={false}
    >
      <EuiButton
        color="primary"
        fill={false}
        iconSide="left"
        onClick={[Function]}
        size="m"
        type="button"
      >
        Run
      </EuiButton>
    </EuiFlexItem>
  </EuiFlexGroup>
</div>
`;
